include/verilated_vpi.cpp

```diff
@@ -2812,13 +2824,1086 @@ vpiHandle vpi_put_value(vpiHandle object, p_vpi_value valuep, p_vpi_time /*time_
+
+    if (arrayvalue_p->format == vpiVectorVal) {
+        switch (varp->vltype()) {
+            case VLVT_UINT8:
+            case VLVT_UINT16:
+            case VLVT_UINT32:
+            case VLVT_UINT64:
+            case VLVT_WDATA:
+                return status;
+            default:
+                status = false;
+        }
+    } else if (arrayvalue_p->format == vpiIntVal) {
+        switch (varp->vltype()) {
+            case VLVT_UINT8:
+            case VLVT_UINT16:
+            case VLVT_UINT32:
+                return status;
+            default:
+                status = false;
+        }
+    } else if (arrayvalue_p->format == vpiRealVal) {
+        switch (varp->vltype()) {
+            case VLVT_UINT64:
+                return status;
+            default:
+                status = false;
+        }
+    } else if ((arrayvalue_p->format == vpiRawTwoStateVal)
+        || (arrayvalue_p->format == vpiRawFourStateVal)) {
+        switch (varp->vltype()) {
+            case VLVT_UINT8:
+            case VLVT_UINT16:
+            case VLVT_UINT32:
+            case VLVT_UINT64:
+            case VLVT_WDATA:
+                return status;
+            default:
+                status = false;
+        }
+    } else if (arrayvalue_p->format == vpiShortIntVal) {
+        switch (varp->vltype()) {
+            case VLVT_UINT8:
+            case VLVT_UINT16:
+                return status;
+            default:
+                status = false;
+        }
+    } else if (arrayvalue_p->format == vpiLongIntVal) {
+        switch (varp->vltype()) {
+            case VLVT_UINT8:
+            case VLVT_UINT16:
+            case VLVT_UINT64:
+                return status;
+            default:
+                status = false;
+        }
+    }
+
+    VL_VPI_ERROR_(__FILE__, __LINE__, "%s: Unsupported format (%s) as requested for %s",
+                __func__, VerilatedVpiError::strFromVpiVal(arrayvalue_p->format), fullname);
+
+    return status;
 }
-void vpi_put_value_array(vpiHandle /*object*/, p_vpi_arrayvalue /*arrayvalue_p*/,
-                         PLI_INT32* /*index_p*/, PLI_UINT32 /*num*/) {
-    VL_VPI_UNIMP_();
+
+void vl_get_value_array(vpiHandle object, p_vpi_arrayvalue arrayvalue_p,
+                        PLI_INT32* index_p, PLI_UINT32 num) {
+    const VerilatedVpioVar* const vop = VerilatedVpioVar::castp(object);
+    if (!vl_check_array_format(vop->varp(), arrayvalue_p, vop->fullname())) return;
+
+    const VerilatedVar* const varp = vop->varp();
+    int size = vop->size();
+    int index = index_p[0];
+
+    static thread_local EData out_ptr[VL_VALUE_STRING_MAX_WORDS * 2];
+
+    if (VL_UNCOVERABLE(num > size)) {
+        VL_VPI_ERROR_(__FILE__, __LINE__, "%s: requested elements (%u) exceed array size (%u)",
+                    __func__, num, size);
+    }
+
+    index -= fmin(vop->rangep()->left(), vop->rangep()->right());
+
+    if (arrayvalue_p->format == vpiIntVal) {
+        PLI_INT32 *integers;
```

Comment:
```suggestion
        PLI_INT32* integersp;
```

And similarly below use "p" to end pointers.